File: bindings/haskell/AuroFamily.hs
{- |
Module      : AuroFamily
Description : AURO atomic-to-frontier family and embedded-council contract.
Copyright   : (c) Alfredo Medina / ItsNotAILABS, 2026
License     : Apache-2.0
Stability   : Experimental architecture contract; no trained checkpoint claim.

Contract version: 2.1.0
-}

{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE StrictData #-}

module AuroFamily
  ( contractVersion
  , familyId
  , specialistTriad
  , ModelTier(..)
  , SubAgentRole(..)
  , tierRank
  , modelIdForTier
  , tierForModelId
  , parameterTarget
  , roleDefaultModel
  , ArchitectureSpec(..)
  , ModelLane(..)
  , SubAgentSpec(..)
  , SubAgentDispatch(..)
  , familyModelIds
  , canHost
  , resolveChildModelId
  , builtinLanes
  ) where

import GHC.Generics (Generic)

contractVersion :: String
contractVersion = "2.1.0"

familyId :: String
familyId = "Auro"

specialistTriad :: [String]
specialistTriad = ["Auro-500M-SENSUS", "Auro-500M-PRAXIS", "Auro-500M-VERBUM"]

data ModelTier = Atomic | Edge | Specialist | General | Orchestrator | Frontier
  deriving (Eq, Ord, Show, Read, Generic, Enum, Bounded)

tierRank :: ModelTier -> Int
tierRank Atomic       = 0
tierRank Edge         = 1
tierRank Specialist   = 2
tierRank General      = 3
tierRank Orchestrator = 4
tierRank Frontier     = 5

modelIdForTier :: ModelTier -> String
modelIdForTier Atomic       = "Auro-500M"
modelIdForTier Edge         = "Auro-2B"
modelIdForTier Specialist   = "Auro-4B"
modelIdForTier General      = "Auro-8B"
modelIdForTier Orchestrator = "Auro-14B"
modelIdForTier Frontier     = "Auro-100B"

tierForModelId :: String -> Maybe ModelTier
tierForModelId "Auro-156K" = Just Atomic
tierForModelId "Auro-250M" = Just Atomic
tierForModelId "Auro-500M" = Just Atomic
tierForModelId "Auro-2B"   = Just Edge
tierForModelId "Auro-4B"   = Just Specialist
tierForModelId "Auro-8B"   = Just General
tierForModelId "Auro-14B"  = Just Orchestrator
tierForModelId "Auro-100B" = Just Frontier
tierForModelId _            = Nothing

parameterTarget :: String -> Maybe Integer
parameterTarget "Auro-156K" = Just 156000
parameterTarget "Auro-250M" = Just 250000000
parameterTarget "Auro-500M" = Just 500000000
parameterTarget "Auro-2B"   = Just 2000000000
parameterTarget "Auro-4B"   = Just 4000000000
parameterTarget "Auro-8B"   = Just 8000000000
parameterTarget "Auro-14B"  = Just 14000000000
parameterTarget "Auro-100B" = Just 100000000000
parameterTarget _            = Nothing

familyModelIds :: [String]
familyModelIds = ["Auro-156K", "Auro-250M", "Auro-500M", "Auro-2B", "Auro-4B", "Auro-8B", "Auro-14B", "Auro-100B"]

data SubAgentRole
  = RoutingSeed | Classifier | JsonRepair | ToolSelection
  | IntentExtract | RetrievalFilter | StructuredTransform | CodeTriage | MemoryConsolidation | SemanticOutline
  | ToolExecutionPlan | CodePatch | EvidenceReview | LocalWorker | ExpertConsensus | TextExpansion
  | Router | ToolCall | EmbedFast | SpectralTriage
  | CodeEdit | SpectralMatch | JsonStruct | ToolPlan
  | Reason | Plan | Critique | SpectralExplain
  | OrchestratorRole | CouncilChair | InstructDev | MultiAgentRouter
  | FrontierResearch | LongHorizon | SafetyReview | DeepCouncil
  deriving (Eq, Ord, Show, Read, Generic, Enum, Bounded)

roleDefaultModel :: SubAgentRole -> String
roleDefaultModel RoutingSeed      = "Auro-156K"
roleDefaultModel Classifier       = "Auro-156K"
roleDefaultModel JsonRepair       = "Auro-156K"
roleDefaultModel ToolSelection    = "Auro-156K"
roleDefaultModel IntentExtract    = "Auro-250M"
roleDefaultModel RetrievalFilter  = "Auro-250M"
roleDefaultModel StructuredTransform = "Auro-250M"
roleDefaultModel CodeTriage       = "Auro-250M"
roleDefaultModel MemoryConsolidation = "Auro-250M"
roleDefaultModel SemanticOutline  = "Auro-250M"
roleDefaultModel ToolExecutionPlan = "Auro-500M"
roleDefaultModel CodePatch        = "Auro-500M"
roleDefaultModel EvidenceReview   = "Auro-500M"
roleDefaultModel LocalWorker      = "Auro-500M"
roleDefaultModel ExpertConsensus  = "Auro-500M"
roleDefaultModel TextExpansion    = "Auro-500M"
roleDefaultModel Router           = "Auro-2B"
roleDefaultModel ToolCall         = "Auro-2B"
roleDefaultModel EmbedFast        = "Auro-2B"
roleDefaultModel SpectralTriage   = "Auro-2B"
roleDefaultModel CodeEdit         = "Auro-4B"
roleDefaultModel SpectralMatch    = "Auro-4B"
roleDefaultModel JsonStruct       = "Auro-4B"
roleDefaultModel ToolPlan         = "Auro-4B"
roleDefaultModel Reason           = "Auro-8B"
roleDefaultModel Plan             = "Auro-8B"
roleDefaultModel Critique         = "Auro-8B"
roleDefaultModel SpectralExplain  = "Auro-8B"
roleDefaultModel OrchestratorRole = "Auro-14B"
roleDefaultModel CouncilChair     = "Auro-14B"
roleDefaultModel InstructDev      = "Auro-14B"
roleDefaultModel MultiAgentRouter = "Auro-14B"
roleDefaultModel FrontierResearch = "Auro-100B"
roleDefaultModel LongHorizon      = "Auro-100B"
roleDefaultModel SafetyReview     = "Auro-100B"
roleDefaultModel DeepCouncil      = "Auro-100B"

data ArchitectureSpec = ArchitectureSpec
  { archHiddenSize                :: !Int
  , archLayers                    :: !Int
  , archAttentionHeads            :: !Int
  , archKvHeads                   :: !Int
  , archIntermediateSize          :: !Int
  , archContextWindowTokensTarget :: !Int
  , archVocabSizeTarget           :: !Int
  } deriving (Eq, Show, Generic)

data ModelLane = ModelLane
  { laneModelId           :: !String
  , laneParameterTarget   :: !Integer
  , laneTier              :: !ModelTier
  , laneCanEmbedSubagents :: !Bool
  , laneEmbeddableTiers   :: ![ModelTier]
  , laneRoles             :: ![SubAgentRole]
  } deriving (Eq, Show, Generic)

data SubAgentSpec = SubAgentSpec
  { specAgentId       :: !String
  , specRole          :: !SubAgentRole
  , specChildModelId  :: !String
  , specParentModelId :: !String
  , specTaskId        :: !String
  , specIntent        :: !String
  , specEvidenceRefs  :: ![String]
  } deriving (Eq, Show, Generic)

data SubAgentDispatch = SubAgentDispatch
  { dispatchOk             :: !Bool
  , dispatchParentModelId  :: !String
  , dispatchChildModelId   :: !String
  , dispatchRole           :: !SubAgentRole
  , dispatchAgentId        :: !String
  , dispatchTaskId         :: !String
  , dispatchMessage        :: !String
  , dispatchError          :: !(Maybe String)
  } deriving (Eq, Show, Generic)

canHost :: ModelTier -> ModelTier -> Bool
canHost parent child = tierRank parent > tierRank child

resolveChildModelId :: String -> SubAgentRole -> Either String String
resolveChildModelId parentModelId role = do
  parentTier <- maybe (Left $ "unknown parent model_id: " ++ parentModelId) Right (tierForModelId parentModelId)
  let child = roleDefaultModel role
  childTier <- maybe (Left $ "unknown child model_id: " ++ child) Right (tierForModelId child)
  if canHost parentTier childTier
    then Right child
    else Left "incompatible parent and child lane"

builtinLanes :: [ModelLane]
builtinLanes =
  [ ModelLane "Auro-156K" 156000 Atomic False [] [RoutingSeed, Classifier, JsonRepair, ToolSelection]
  , ModelLane "Auro-250M" 250000000 Atomic False [] [IntentExtract, RetrievalFilter, StructuredTransform, CodeTriage, MemoryConsolidation, SemanticOutline]
  , ModelLane "Auro-500M" 500000000 Atomic False [] [ToolExecutionPlan, CodePatch, EvidenceReview, LocalWorker, ExpertConsensus, TextExpansion]
  , ModelLane "Auro-2B" 2000000000 Edge True [Atomic] [Router, ToolCall, EmbedFast, SpectralTriage]
  , ModelLane "Auro-4B" 4000000000 Specialist True [Atomic, Edge] [CodeEdit, SpectralMatch, JsonStruct, ToolPlan]
  , ModelLane "Auro-8B" 8000000000 General True [Atomic, Edge, Specialist] [Reason, Plan, Critique, SpectralExplain]
  , ModelLane "Auro-14B" 14000000000 Orchestrator True [Atomic, Edge, Specialist, General] [OrchestratorRole, CouncilChair, InstructDev, MultiAgentRouter]
  , ModelLane "Auro-100B" 100000000000 Frontier True [Atomic, Edge, Specialist, General, Orchestrator] [FrontierResearch, LongHorizon, SafetyReview, DeepCouncil]
  ]

File: bindings/julia/AuroFamily/src/AuroFamily.jl
"""
Julia mirror of the AURO family contract.
Contract version 2.1.0. Architecture and composition only; no checkpoint claim.
"""
module AuroFamily

export CONTRACT_VERSION, FAMILY_ID, FAMILY_MODEL_IDS, SPECIALIST_TRIAD
export ModelTier, SubAgentRole, ModelLane
export model_id_for_tier, tier_for_model_id, parameter_target
export can_host, resolve_child_model_id, family_model_ids, builtin_lanes

const CONTRACT_VERSION = "2.1.0"
const FAMILY_ID = "Auro"

@enum ModelTier begin
    ATOMIC
    EDGE
    SPECIALIST
    GENERAL
    ORCHESTRATOR
    FRONTIER
end

@enum SubAgentRole begin
    ROUTING_SEED
    CLASSIFIER
    JSON_REPAIR
    TOOL_SELECTION
    INTENT_EXTRACT
    RETRIEVAL_FILTER
    STRUCTURED_TRANSFORM
    CODE_TRIAGE
    MEMORY_CONSOLIDATION
    SEMANTIC_OUTLINE
    TOOL_EXECUTION_PLAN
    CODE_PATCH
    EVIDENCE_REVIEW
    LOCAL_WORKER
    EXPERT_CONSENSUS
    TEXT_EXPANSION
    ROUTER
    TOOL_CALL
    EMBED_FAST
    SPECTRAL_TRIAGE
    CODE_EDIT
    SPECTRAL_MATCH
    JSON_STRUCT
    TOOL_PLAN
    REASON
    PLAN
    CRITIQUE
    SPECTRAL_EXPLAIN
    ORCHESTRATOR_ROLE
    COUNCIL_CHAIR
    INSTRUCT_DEV
    MULTI_AGENT_ROUTER
    FRONTIER_RESEARCH
    LONG_HORIZON
    SAFETY_REVIEW
    DEEP_COUNCIL
end

struct ModelLane
    model_id::String
    parameter_target::Int
    tier::ModelTier
    can_embed_subagents::Bool
    embeddable_tiers::Vector{ModelTier}
    roles::Vector{SubAgentRole}
end

const TIER_RANK = Dict(ATOMIC=>0, EDGE=>1, SPECIALIST=>2, GENERAL=>3, ORCHESTRATOR=>4, FRONTIER=>5)
const TIER_TO_MODEL = Dict(ATOMIC=>"Auro-500M", EDGE=>"Auro-2B", SPECIALIST=>"Auro-4B", GENERAL=>"Auro-8B", ORCHESTRATOR=>"Auro-14B", FRONTIER=>"Auro-100B")
const FAMILY_MODEL_IDS = ["Auro-156K", "Auro-250M", "Auro-500M", "Auro-2B", "Auro-4B", "Auro-8B", "Auro-14B", "Auro-100B"]
const MODEL_TO_TIER = Dict("Auro-156K"=>ATOMIC, "Auro-250M"=>ATOMIC, "Auro-500M"=>ATOMIC, "Auro-2B"=>EDGE, "Auro-4B"=>SPECIALIST, "Auro-8B"=>GENERAL, "Auro-14B"=>ORCHESTRATOR, "Auro-100B"=>FRONTIER)
const PARAM_TARGETS = Dict("Auro-156K"=>156_000, "Auro-250M"=>250_000_000, "Auro-500M"=>500_000_000, "Auro-2B"=>2_000_000_000, "Auro-4B"=>4_000_000_000, "Auro-8B"=>8_000_000_000, "Auro-14B"=>14_000_000_000, "Auro-100B"=>100_000_000_000)
const SPECIALIST_TRIAD = ["Auro-500M-SENSUS", "Auro-500M-PRAXIS", "Auro-500M-VERBUM"]

const ROLE_DEFAULT_MODEL = Dict(
    ROUTING_SEED=>"Auro-156K", CLASSIFIER=>"Auro-156K", JSON_REPAIR=>"Auro-156K", TOOL_SELECTION=>"Auro-156K",
    INTENT_EXTRACT=>"Auro-250M", RETRIEVAL_FILTER=>"Auro-250M", STRUCTURED_TRANSFORM=>"Auro-250M", CODE_TRIAGE=>"Auro-250M", MEMORY_CONSOLIDATION=>"Auro-250M", SEMANTIC_OUTLINE=>"Auro-250M",
    TOOL_EXECUTION_PLAN=>"Auro-500M", CODE_PATCH=>"Auro-500M", EVIDENCE_REVIEW=>"Auro-500M", LOCAL_WORKER=>"Auro-500M", EXPERT_CONSENSUS=>"Auro-500M", TEXT_EXPANSION=>"Auro-500M",
    ROUTER=>"Auro-2B", TOOL_CALL=>"Auro-2B", EMBED_FAST=>"Auro-2B", SPECTRAL_TRIAGE=>"Auro-2B",
    CODE_EDIT=>"Auro-4B", SPECTRAL_MATCH=>"Auro-4B", JSON_STRUCT=>"Auro-4B", TOOL_PLAN=>"Auro-4B",
    REASON=>"Auro-8B", PLAN=>"Auro-8B", CRITIQUE=>"Auro-8B", SPECTRAL_EXPLAIN=>"Auro-8B",
    ORCHESTRATOR_ROLE=>"Auro-14B", COUNCIL_CHAIR=>"Auro-14B", INSTRUCT_DEV=>"Auro-14B", MULTI_AGENT_ROUTER=>"Auro-14B",
    FRONTIER_RESEARCH=>"Auro-100B", LONG_HORIZON=>"Auro-100B", SAFETY_REVIEW=>"Auro-100B", DEEP_COUNCIL=>"Auro-100B"
)

model_id_for_tier(t::ModelTier) = TIER_TO_MODEL[t]
tier_for_model_id(id::AbstractString) = MODEL_TO_TIER[String(id)]
parameter_target(id::AbstractString) = PARAM_TARGETS[String(id)]
family_model_ids() = copy(FAMILY_MODEL_IDS)
can_host(parent_tier::ModelTier, child_tier::ModelTier) = TIER_RANK[parent_tier] > TIER_RANK[child_tier]

function resolve_child_model_id(parent_model_id::AbstractString, role::SubAgentRole)::String
    child = ROLE_DEFAULT_MODEL[role]
    can_host(tier_for_model_id(parent_model_id), tier_for_model_id(child)) || error("incompatible parent and child lane")
    return child
end

function builtin_lanes()::Vector{ModelLane}
    return [
        ModelLane("Auro-156K", PARAM_TARGETS["Auro-156K"], ATOMIC, false, ModelTier[], [ROUTING_SEED, CLASSIFIER, JSON_REPAIR, TOOL_SELECTION]),
        ModelLane("Auro-250M", PARAM_TARGETS["Auro-250M"], ATOMIC, false, ModelTier[], [INTENT_EXTRACT, RETRIEVAL_FILTER, STRUCTURED_TRANSFORM, CODE_TRIAGE, MEMORY_CONSOLIDATION, SEMANTIC_OUTLINE]),
        ModelLane("Auro-500M", PARAM_TARGETS["Auro-500M"], ATOMIC, false, ModelTier[], [TOOL_EXECUTION_PLAN, CODE_PATCH, EVIDENCE_REVIEW, LOCAL_WORKER, EXPERT_CONSENSUS, TEXT_EXPANSION]),
        ModelLane("Auro-2B", PARAM_TARGETS["Auro-2B"], EDGE, true, [ATOMIC], [ROUTER, TOOL_CALL, EMBED_FAST, SPECTRAL_TRIAGE]),
        ModelLane("Auro-4B", PARAM_TARGETS["Auro-4B"], SPECIALIST, true, [ATOMIC, EDGE], [CODE_EDIT, SPECTRAL_MATCH, JSON_STRUCT, TOOL_PLAN]),
        ModelLane("Auro-8B", PARAM_TARGETS["Auro-8B"], GENERAL, true, [ATOMIC, EDGE, SPECIALIST], [REASON, PLAN, CRITIQUE, SPECTRAL_EXPLAIN]),
        ModelLane("Auro-14B", PARAM_TARGETS["Auro-14B"], ORCHESTRATOR, true, [ATOMIC, EDGE, SPECIALIST, GENERAL], [ORCHESTRATOR_ROLE, COUNCIL_CHAIR, INSTRUCT_DEV, MULTI_AGENT_ROUTER]),
        ModelLane("Auro-100B", PARAM_TARGETS["Auro-100B"], FRONTIER, true, [ATOMIC, EDGE, SPECIALIST, GENERAL, ORCHESTRATOR], [FRONTIER_RESEARCH, LONG_HORIZON, SAFETY_REVIEW, DEEP_COUNCIL])
    ]
end

end
